// metro-visualizer/src/app/graph/components/SearchBar.js
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 * @format
 */

/* eslint-env browser */

'use strict';

const FuzzySearch = require('fuzzy-search');
const React = require('react');var _require =

require('emotion');const css = _require.css;

import { Icon, Input, AutoComplete } from 'antd';











class SearchBar extends React.Component {



  constructor(props) {
    super(props);this.state = { results: [], query: '' };this.



    handleSearch = value => {
      this.setState({
        results: this.searcher.search(value),
        query: value });

    };this.

    handleSubmit = value => {
      this.setState({ query: value.substring(value.lastIndexOf('/') + 1) });
      this.props.onSelection(value);
    };this.searcher = new FuzzySearch(props.data, ['name'], { sort: true });}

  render() {
    return (
      React.createElement(AutoComplete, {
          size: 'large',
          value: this.state.query,
          className: searchBar,
          onSearch: this.handleSearch,
          onSelect: this.handleSubmit,
          placeholder: 'Search for modules',
          dataSource: this.state.results.slice(0, 10).map(module =>
          React.createElement(AutoComplete.Option, { key: module.filePath, value: module.filePath },
            module.name,
            React.createElement('p', { style: { fontSize: '7pt', marginBottom: '0px' } },
              module.filePath))) },



        React.createElement(Input, {
          suffix: React.createElement(Icon, { type: 'search', className: 'certain-category-icon' }) })));



  }}


const searchBar = css`
  margin: 0px 8px 2px 8px;
  width: 100%;
  font-size: 11pt;
`;

module.exports = SearchBar;
